### PR TITLE
Problem 2: Remove Zero Sum Consecutive Nodes

### PR DESCRIPTION
**Pull request by mduw**

### CODE REVIEW

File: 2020-02-Spring/2_linked_lists/README.md [hunk 1]
```diff
@@ -45,22 +45,43 @@ Go to [Solution](#s1)   [Top](#top)
 <!-- Don't remove -->
 <a name="p2"/>
 
-### 2. PROBLEM 2 TODO :bug:
+### 2. Remove Zero Sum Consecutive Nodes from Linked List
 
-Source: TODO :bug:
+Source: https://leetcode.com/problems/remove-zero-sum-consecutive-nodes-from-linked-list/
```
Comment (anishprasad01) [inline 2020-02-Spring/2_linked_lists/README.md:50]:
```suggestion
Source: Leetcode[https://leetcode.com/problems/remove-zero-sum-consecutive-nodes-from-linked-list/]
```

File: 2020-02-Spring/2_linked_lists/README.md [hunk 2]
```diff
@@ -45,22 +45,43 @@ Go to [Solution](#s1)   [Top](#top)
...
 
 #### Scenario
 
-Problem Statement TODO :bug:
+Given the head of a linked list, we repeatedly delete consecutive sequences of nodes that sum to 0 until there are no such sequences.
 
-#### Example Input
+After doing so, return the head of the final linked list.  You may return any such answer.
```
Comment (anishprasad01) [inline 2020-02-Spring/2_linked_lists/README.md:56]: Clarify what "any such answer" means

File: 2020-02-Spring/2_linked_lists/README.md [hunk 3]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
 <!-- Don't remove -->
 <a name="s2"/>
 
-### 2. SOLUTION 2 TODO :bug:
-
-Source: TODO :bug:
+### 2. Remove Zero Sum Consecutive Nodes from Linked List
 
 #### Naive/Simple Solution
+The idea is to locate the sublist of nodes that sum to 0 and de-list them.<br>
```
Comment (anishprasad01) [inline 2020-02-Spring/2_linked_lists/README.md:153]: I personally found this explanation hard to follow. Please consider revising it

File: 2020-02-Spring/2_linked_lists/README.md [hunk 4]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
+
+###### The following diagram illustrates how the algorithm works.
+
+![Algorithm Illustration](./ZeroSumConsec/Algorithm_Illustration.png)
 
 <details>
 <summary>Click to see solution</summary>
 
-TODO put your solution here :bug:
+```
```
Comment (anishprasad01) [inline 2020-02-Spring/2_linked_lists/README.md:215]:
```suggestion
```C++
```

File: 2020-02-Spring/2_linked_lists/README.md [hunk 5]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
 </details>
 
 #### Optimal Solution
+One approach is to calculate the accumulated sum that is defined by Sum<sub>i</sub> = Sum <sub>i - 1</sub> + List <sub>i</sub> -> val. The accumulated Sum is tracked by using the hash table SumMap.
+- if Sum is found in SumMap (accumulated Sum is repeated), this means there is a zero-sum range existing between the two nodes, then we remove all nodes between those two(s) and continue scanning until no zero-sum range found.
+- if Sum is not repeated, this means there is NO zero-sum range existing between the two nodes, therefore, we will record the accumulated sum and examine the next node. <br>
+- Instead of going through from (N-i) nodes for every i-th node explained in the previous solution, this solution uses a hash table to record the node before the zero-sum range and only goes through the list once.
+- Time complexity of this solution is O(N). <br>
+
+###### The following diagram illustrates how the algorithm works.
```
Comment (anishprasad01) [inline 2020-02-Spring/2_linked_lists/README.md:208]: Consider hiding this so people working on Q1 don't accidentally see it.

File: 2020-02-Spring/2_linked_lists/README.md (hunk 1 above)
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:50]: Use markdown link embedding syntax.

```suggestion
Source: [LeetCode](https://leetcode.com/problems/remove-zero-sum-consecutive-nodes-from-linked-list/)
```

File: 2020-02-Spring/2_linked_lists/README.md [hunk 6]
```diff
@@ -45,22 +45,43 @@ Go to [Solution](#s1)   [Top](#top)
...
+Given the head of a linked list, we repeatedly delete consecutive sequences of nodes that sum to 0 until there are no such sequences.
 
-#### Example Input
+After doing so, return the head of the final linked list.  You may return any such answer.
 
-If the problem is simple enough, remove this section. TODO :bug:
+(Note that in the examples below, all sequences are serializations of ListNode objects.)
 
+#### Example Input
+##### Example 1: <br>
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:61]: the HTML `<br>` is fine, but shouldn't be necessary. When in doubt, include whitespace between paragraphs and headings instead of using HTML.

File: 2020-02-Spring/2_linked_lists/README.md [hunk 7]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
-Source: TODO :bug:
+### 2. Remove Zero Sum Consecutive Nodes from Linked List
 
 #### Naive/Simple Solution
+The idea is to locate the sublist of nodes that sum to 0 and de-list them.<br>
+- Starting at the edging node, find lower-bound and upper-bound node of the sublist that sum to 0.
+- if found, de-list nodes from [LBound, UBound], including by re-link PreLBound->next = UBound->next;
+- If not found, move forward and check the next node.
+For every i-th node of N nodes, the algorithm iterates from node i-th to the last node for the total of N*(N-i) times.
+- This solution has time complexity of O(N<sup>2</sup>).
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:158]: Recommend using monospacing for all complexity estimations and inline pseudocode.

```suggestion
- This solution has time complexity of `O(N^2)`.
```

File: 2020-02-Spring/2_linked_lists/README.md [hunk 8]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
 
-### 2. SOLUTION 2 TODO :bug:
-
-Source: TODO :bug:
+### 2. Remove Zero Sum Consecutive Nodes from Linked List
 
 #### Naive/Simple Solution
+The idea is to locate the sublist of nodes that sum to 0 and de-list them.<br>
+- Starting at the edging node, find lower-bound and upper-bound node of the sublist that sum to 0.
+- if found, de-list nodes from [LBound, UBound], including by re-link PreLBound->next = UBound->next;
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:155]: A few suggestions:

* Recommend using monospacing for all complexity estimations and inline pseudocode. 
* Be consistent with capitalizing the start of a bullet -- either do it always or never do it.
* Explicitly mention that the bounds are inclusive; _I_ realize that you're using that notation to indicate inclusive bounds, but not everyone will understand that.
* Use phrasing which is more technical than "de-list" and "including by re-link".
* Make the second half of your sentence a complete clause.

```suggestion
- If found, remove nodes from `[LBound, UBound]` inclusive, and be sure to reset the preceding node's pointer, e.g. `PreLBound->next = UBound->next`;
```

File: 2020-02-Spring/2_linked_lists/README.md (hunk 3 above)
Comment (etcadinfinitum) [reply 2020-02-Spring/2_linked_lists/README.md:153]: I second this. The mechanics of deleting nodes and resetting list pointers appropriately seem fine, but your explanation glosses over how you actually find the zero-sum sublist. It is _this_ component which we want to hear about, not necessarily Linked List manipulation mechanics.

File: 2020-02-Spring/2_linked_lists/README.md [hunk 9]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
+- If not found, move forward and check the next node.
+For every i-th node of N nodes, the algorithm iterates from node i-th to the last node for the total of N*(N-i) times.
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:157]: This bullet does not explain the relationship between inner loop complexity to search for a zero sum (time complexity `O(N - i)`) and the outer loop (`O(N)`). Recommend removing this bullet.
```suggestion
```

File: 2020-02-Spring/2_linked_lists/README.md [hunk 10]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
-### 2. SOLUTION 2 TODO :bug:
-
-Source: TODO :bug:
+### 2. Remove Zero Sum Consecutive Nodes from Linked List
 
 #### Naive/Simple Solution
+The idea is to locate the sublist of nodes that sum to 0 and de-list them.<br>
+- Starting at the edging node, find lower-bound and upper-bound node of the sublist that sum to 0.
+- if found, de-list nodes from [LBound, UBound], including by re-link PreLBound->next = UBound->next;
+- If not found, move forward and check the next node.
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:156]: Are we checking the next node to see if it contributes to our zero sum, or are we starting a new outer loop iteration? Revise the phrasing here.
```suggestion
- Upon reaching the end of the list in this iteration of zero-sum searching, repeat the process beginning at the next node in the list.
```

File: 2020-02-Spring/2_linked_lists/README.md [hunk 11]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
+```
 
 </details>
 
 #### Optimal Solution
+One approach is to calculate the accumulated sum that is defined by Sum<sub>i</sub> = Sum <sub>i - 1</sub> + List <sub>i</sub> -> val. The accumulated Sum is tracked by using the hash table SumMap.
+- if Sum is found in SumMap (accumulated Sum is repeated), this means there is a zero-sum range existing between the two nodes, then we remove all nodes between those two(s) and continue scanning until no zero-sum range found.
+- if Sum is not repeated, this means there is NO zero-sum range existing between the two nodes, therefore, we will record the accumulated sum and examine the next node. <br>
+- Instead of going through from (N-i) nodes for every i-th node explained in the previous solution, this solution uses a hash table to record the node before the zero-sum range and only goes through the list once.
+- Time complexity of this solution is O(N). <br>
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:206]: A few suggestions:

1. I recommend using monospace for all code-like snippets you're including here. The subscripts are neat, but inline code snippets draws a better distinction for variable names.
2. No `<br>` elements are required in an unordered list (which is what the `- ` prefix does before each line.
3. This is a good solution, which optimizes for time complexity, **but takes your space complexity from `O(1)` to `O(N)`** because you are allocating additional space, potentially one extra map entry for every element in your list. You should mention this tradeoff.

File: 2020-02-Spring/2_linked_lists/README.md [hunk 12]
```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
...
+
+###### The following diagram illustrates how the algorithm works.
+
+![Algorithm Illustration](./ZeroSumConsec/Algorithm_Illustration.png)
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:210]: I approve of this diagram :+1:

File: 2020-02-Spring/2_linked_lists/README.md [hunk 13]
```diff
@@ -45,22 +45,43 @@ Go to [Solution](#s1)   [Top](#top)
...
+```
+Input: head = [5,-3,-4,1,6,-2,5]
+Output: [5,-2,5]
+```
 #### Function Signature
-
-TODO :bug:
-
+##### C++:
+```
```
Comment (etcadinfinitum) [inline 2020-02-Spring/2_linked_lists/README.md:73]:
````suggestion
```c++
````

File: 2020-02-Spring/2_linked_lists/README.md (hunk 2 above)
Comment (etcadinfinitum) [reply 2020-02-Spring/2_linked_lists/README.md:56]: I agree. Clarify what this means, or remove the second sentence completely.

If you are referring to a `nullptr` head, I wouldn't try to drop hints about that. The participants should be asking clarifying questions about that if it doesn't make sense.

```suggestion
After doing so, return the head of the final linked list.
```

File: 2020-02-Spring/2_linked_lists/README.md (hunk 1 above)
Comment (anishprasad01) [reply 2020-02-Spring/2_linked_lists/README.md:50]: Whoops, my bad.

File: 2020-02-Spring/2_linked_lists/README.md [hunk 14]
```diff
@@ -126,31 +146,125 @@ Go to [Top](#top)
...
 
 #### Naive/Simple Solution
 
 <details>
 <summary>Click to see solution</summary>
 
-TODO put your solution here :bug:
+// NOTE: This solution does not take into account memory leaks. <br>
+The idea is to use an outer loop iterating through N nodes, at every node i-th we use a nested loop iterating to the end of the list to check whether or not that node and the following nodes cause a zero-sum. The list of nodes causes zero-sum if "the accumulated sum of all nodes in that list equals zero". The inner loop stops as the accumulated sum equals 0 ``Sum == 0`` or it reaches the end of the list.
+- If zero-sum list is found by the inner loop ``Sum == 0``, we de-list nodes from nodes [LBound, UBound] including the LBound node and the UBound node, where the node UBound is the node that the inner loop stopped at as the ``Sum == 0``.
```
Comment (anishprasad01) [inline 2020-02-Spring/2_linked_lists/README.md:158]:
```suggestion
- If zero-sum list is found by the inner loop ``Sum == 0``, we remove nodes between [LBound, UBound] including the LBound node and the UBound node, where the node UBound is the node that the inner loop stopped at as the ``Sum == 0``.
```
